File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        
        charSet = set()
        
        a = 0
        b = 0
        MAX = 0
        while a < len(s):
            if s[a] not in charSet:
                charSet.add(s[a])
                a += 1
                MAX = max(MAX, len(charSet))
            else:
                charSet.remove(s[b])
                b += 1
            
        return MAX
                

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        if (s.length() == 0) return 0;
        HashMap<Character, Integer> lastSeen = new HashMap<>();
        int l = 0, longest = 0;
        for (int r = 0; r < s.length(); r++) {
            char ch = s.charAt(r);
            if (lastSeen.containsKey(ch)) {
                l = Math.max(l, lastSeen.get(ch)+1);
            }
            lastSeen.put(ch, r);
            longest = Math.max(longest, (r-l+1));
        }
        return longest;
    }
}